Add golangci-lint to golang-1.24 image

czeslavo opened the pull request:
That's a proposal to add golangci-lint to the golang image so we can use it in CI.

rzetelskik commented on golang/1.24/versions.sh:
```diff
@@ -0,0 +1,2 @@
+GO_VERSION=1.24.3
+GOLANGCI_LINT_VERSION=2.1.6
```
this will break the update logic https://github.com/scylladb/scylla-operator-images/blob/69741464fd17785b3b9a3508270ca0e5b97c06e7/golang/update.sh#L22, I'm not sure GOLANGCI_LINT_VERSION is needed though

rzetelskik commented on golang/1.24/lib.sh:
```diff
@@ -22,3 +22,13 @@ function install-golang {
+  version="${1}"
+  url="https://github.com/golangci/golangci-lint/releases/download/v${version}/golangci-lint-${version}-linux-${2}.tar.gz"
+  f="$( basename "${url}" )"
+  curl --fail --retry 5 --retry-all-errors -L "${url}" > "/tmp/${f}"
```
The whole containerfile is using /tmp/ explicitly. It should be changed on the upper level if you think it should be changed

rzetelskik commented on golang/1.24/lib.sh:
```diff
@@ -22,3 +22,13 @@ function install-golang {
+  version="${1}"
+  url="https://github.com/golangci/golangci-lint/releases/download/v${version}/golangci-lint-${version}-linux-${2}.tar.gz"
+  f="$( basename "${url}" )"
+  curl --fail --retry 5 --retry-all-errors -L "${url}" > "/tmp/${f}"
```
let's verify the checksum as well

czeslavo commented:
@rzetelskik I've added proper golangci-lint archive checksum verification and dropped golang-1.23 from the build process as we no longer use 1.23 in our products, as we discussed. I also moved checksums to `golang/{version}/checksums` directory to avoid unnecessary pollution and adjusted scripts to use `mktemp` instead of hardcoding `/tmp`.

zimnx commented:
>  dropped golang-1.23 from the build process as we no longer use 1.23 in our products, as we discussed

We do use golang-1.23 in supported Operator version: https://github.com/scylladb/scylla-operator/blob/v1.16/Dockerfile
Even if we wouldn't have any usage, drop should be part of separate PR. Keep your PR changes scoped.

czeslavo commented:
> We do use golang-1.23 in supported Operator version: https://github.com/scylladb/scylla-operator/blob/v1.16/Dockerfile Even if we wouldn't have any usage, drop should be part of separate PR. Keep your PR changes scoped.

Good catch, thanks. :+1: I agree - reverted the drop and aligned 1.23 to also include golangci-lint.

rzetelskik commented on golang/update.sh:
```diff
@@ -3,23 +3,48 @@
   short_version="${version%.*}"
-  find "${short_version}" -name '*.sum' -delete
+
+  # Resolve golangci-lint version.
+  golangci_lint_version="$( curl --fail --retry 5 --retry-all-errors -L https://api.github.com/repos/golangci/golangci-lint/releases/latest | jq -r '.tag_name' )"
```
why tie this together with the golang version update? Is there any relation between the two?

czeslavo replied:
Tied them because the images are grouped by Golang version. For each Golang version, we have to keep the checksum of golangci-lint that it installs. It also makes sense to have some automation to automatically bump the golangci-lint version (to avoid manual version bump and checksum save), and this covers that. If you think this should be somehow separated, I'm open to suggestions.

In general, golangci-lint is compatible with Golang versions lower or equal to the one it was built with. It means that once a breaking change to some Golang API is introduced, golangci-lint has to adapt.

czeslavo commented on golang/update.sh:
```diff
@@ -3,23 +3,48 @@
+  echo "${1}" | jq -r '.[] | select( .version | test("'"${2}"'") ) | .version | sub("^go"; "")'
+}
+
+function dump-versions {
```
[Addressed](https://github.com/scylladb/scylla-operator-images/compare/64c387f312fe990bc2e4dd158f9a9ecb6ca49f7f..3f466b57734194407c29b331a47edb8780e509d4).
